File: apps/mobile/lib/src/features/stats/data/stats_api_client.dart
import 'dart:convert';

import 'package:dio/dio.dart';

import '../../../core/network/api_client.dart';
import '../../../core/network/api_envelope.dart';
import '../../../core/network/api_error.dart';
import '../domain/local_study_daily_stat.dart';

class StatsApiClient {
  StatsApiClient({Dio? dio, ApiClient? apiClient})
    : _dio = dio ?? apiClient?.dio ?? ApiClient().dio;

  final Dio _dio;

  Future<void> addDailyStats({
    required DateTime statDate,
    required int readingMinutes,
    required int lookupCount,
    required int paragraphTranslationCount,
    required int cardsCreated,
    required int cardsReviewed,
  }) async {
    final response = await _dio.request<Object?>(
      '/api/v1/stats/daily',
      data: {
        'statDate': LocalStudyDailyStat.dateString(statDate),
        'readingMinutes': readingMinutes,
        'lookupCount': lookupCount,
        'paragraphTranslationCount': paragraphTranslationCount,
        'cardsCreated': cardsCreated,
        'cardsReviewed': cardsReviewed,
      },
      options: Options(method: 'POST', validateStatus: (_) => true),
    );
    final envelope = ApiEnvelope.fromJson(_decode(response.data));
    if (!envelope.success) {
      throw envelope.error ??
          const ApiError(code: 'UNKNOWN_ERROR', message: 'Request failed');
    }
  }

  Object? _decode(Object? data) {
    if (data is String) {
      return jsonDecode(data);
    }
    return data;
  }
}

File: apps/web-reader/services/statsApiClient.ts
import { ofetch } from 'ofetch'

import type { ApiEnvelope } from '../types/api'
import { WebApiError } from '../types/api'
import type { WebStudyDailyStatsCounters } from '../types/localData'

type Fetcher = (url: string, options: Record<string, unknown>) => Promise<unknown>

export interface StatsApiClientOptions {
  baseUrl?: string
  fetcher?: Fetcher
}

export interface AddDailyStatsRequest extends WebStudyDailyStatsCounters {
  statDate: string
}

export interface StatsApiClient {
  addDailyStats(input: AddDailyStatsRequest): Promise<unknown>
}

export function createStatsApiClient(options: StatsApiClientOptions = {}): StatsApiClient {
  const baseUrl = options.baseUrl ?? ''
  const fetcher = options.fetcher ?? ofetch
  return {
    addDailyStats(input) {
      return request(fetcher, `${baseUrl}/api/v1/stats/daily`, {
        method: 'POST',
        body: {
          statDate: input.statDate,
          readingMinutes: input.readingMinutes,
          lookupCount: input.lookupCount,
          paragraphTranslationCount: input.paragraphTranslationCount,
          cardsCreated: input.cardsCreated,
          cardsReviewed: input.cardsReviewed,
        },
      })
    },
  }
}

async function request(
  fetcher: Fetcher,
  url: string,
  options: Record<string, unknown>,
): Promise<unknown> {
  const envelope = await fetcher(url, options) as ApiEnvelope<unknown>
  if (!envelope.success) {
    const error = envelope.error ?? { code: 'UNKNOWN_ERROR', message: 'Request failed' }
    throw new WebApiError(error.code, error.message)
  }
  return envelope.data
}
